File: examples/cpp/run_async_model_conf/run_async_model_conf.cpp
/*
 * Copyright (C) 2018- DEEPX Ltd.
 * All rights reserved.
 *
 * This software is the property of DEEPX and is provided exclusively to customers
 * who are supplied with DEEPX NPU (Neural Processing Unit).
 * Unauthorized sharing or usage is strictly prohibited by law.
 */

#include "dxrt/dxrt_api.h"
#include "dxrt/extern/cxxopts.hpp"
#include "dxrt/device_info_status.h"
#include "../include/logger.h"

#include <string>
#include <iostream>
#include <condition_variable>


int main(int argc, char* argv[])
{
    std::string model_path;
    int loop_count;
    bool verbose;

    auto& log = dxrt::Logger::GetInstance();

    cxxopts::Options options("run_async_model_conf", "Run asynchronous model inference with configuration");
    options.add_options()
        ("m,model", "Path to model file (.dxnn)", cxxopts::value<std::string>(model_path))
        ("l,loops", "Number of inference loops", cxxopts::value<int>(loop_count)->default_value("1"))
        ("v,verbose", "Enable verbose/debug logging", cxxopts::value<bool>(verbose)->default_value("false"))
        ("h,help", "Print usage");

    try
    {
        auto result = options.parse(argc, argv);

        if (result.count("help") || !result.count("model"))
        {
            std::cout << options.help() << std::endl;
            return result.count("help") ? 0 : -1;
        }

        if (verbose) {
            log.SetLevel(dxrt::Logger::Level::LOGLEVEL_DEBUG);
        }
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("Error parsing arguments: ") + e.what());
        std::cout << options.help() << std::endl;
        return -1;
    }

    log.Info("Start async_model_conf test for model: " + model_path);

    int callback_count = 0;

    try
    {

        // enable show model info and profile
        dxrt::Configuration::GetInstance().SetEnable(dxrt::Configuration::ITEM::SHOW_MODEL_INFO, true);
        dxrt::Configuration::GetInstance().SetEnable(dxrt::Configuration::ITEM::SHOW_PROFILE, true);


        std::mutex cv_mutex;
        std::condition_variable cv;

        // create inference engine instance with model
        dxrt::InferenceEngine ie(model_path);

        // register call back function
        ie.RegisterCallback([&callback_count, &loop_count, &cv_mutex, &cv]
            (const dxrt::TensorPtrs &outputs, const void *userArg) {

            std::ignore = outputs;
            std::ignore = userArg;

            std::unique_lock<std::mutex> lock(cv_mutex);
            callback_count++;
            if ( callback_count == loop_count ) cv.notify_one();

            return 0;
        });

        // create temporary input buffer for example
        std::vector<uint8_t> inputPtr(ie.GetInputSize(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        // inference loop
        for(int i = 0; i < loop_count; ++i)
        {
            // user argument
            auto* userData = new std::pair<int, int>(i, loop_count);

            // inference asynchronously, use all npu cores
            ie.RunAsync(inputPtr.data(), userData);

            log.Debug("Inference request submitted with user_arg(" + std::to_string(i) + ")");

        }

        // wait until all callbacks have been processed
        std::unique_lock<std::mutex> lock(cv_mutex);
        cv.wait(lock, [&callback_count, &loop_count] {
            return callback_count == loop_count;
        });

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count);
        double fps = 1000.0 / avg_latency;

        log.Info("-----------------------------------");
        log.Info("Total Time: " + std::to_string(total_time) + " ms");
        log.Info("Average Latency: " + std::to_string(avg_latency) + " ms");
        log.Info("FPS: " + std::to_string(fps) + " frame/sec");
        log.Info("Total callback-count / loop-count: " +
            std::to_string(callback_count) + " / " + std::to_string(loop_count) +
            (callback_count == loop_count ? " (Success)" : " (Failure)"));
        log.Info("-----------------------------------");


        // device information for each device
        auto device_count = dxrt::DeviceStatus::GetDeviceCount();
        for(int i = 0; i < device_count; ++i)
        {
            auto device_status = dxrt::DeviceStatus::GetCurrentStatus(i);
            log.Info("Device: " + std::to_string(device_status.GetId()));
            log.Info("   Temperature: " + std::to_string(device_status.GetTemperature(0)));
            log.Info("   Voltage: " + std::to_string(device_status.GetNpuVoltage(0)));
            log.Info("   Clock: " + std::to_string(device_status.GetNpuClock(0)));
        }

    }
    catch (const dxrt::Exception& e)
    {
        log.Error(std::string("dxrt::Exception: ") + e.what());
        return -1;
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("std::exception: ") + e.what());
        return -1;
    }
    catch(...)
    {
        log.Error("Exception");
        return -1;
    }

    return (callback_count == loop_count ? 0 : -1);
}

File: examples/python/run_async_model_conf.py
#
# Copyright (C) 2018- DEEPX Ltd.
# All rights reserved.
#
# This software is the property of DEEPX and is provided exclusively to customers
# who are supplied with DEEPX NPU (Neural Processing Unit).
# Unauthorized sharing or usage is strictly prohibited by law.
#

import numpy as np
import argparse
import os
import time
from dx_engine import InferenceEngine, Configuration, DeviceStatus
from logger import Logger, LogLevel

import threading
import queue
from threading import Thread

q = queue.Queue()
gLoopCount = 0

lock = threading.Lock()

def on_inference_callback_func(outputs, user_arg):
    # the outputs are guaranteed to be valid only within this callback function
    # processing this callback functions as quickly as possible is beneficial
    # for improving inference performance
    logger = Logger()
    global gLoopCount

    # Mutex locks should be properly adjusted
    # to ensure that callback functions are thread-safe.
    with lock:

        # user data type casting
        index, loop_count = user_arg

        # post processing
        #postProcessing(outputs);

        # something to do

        logger.debug(f"Inference output (callback) index={index}")

        gLoopCount += 1
        if ( gLoopCount == loop_count ) :
            logger.debug("Complete Callback")
            q.put(0)

    return 0


def parse_args():
    parser = argparse.ArgumentParser(description="Run asynchronous model inference with configuration")
    parser.add_argument("--model", "-m", type=str, required=True, help="Path to model file (.dxnn)")
    parser.add_argument("--loops", "-l", type=int, default=1, help="Number of inference loops (default: 1)")
    parser.add_argument("--verbose", "-v", action="store_true", default=False, help="Enable debug logging")
    args = parser.parse_args()

    if not os.path.exists(args.model):
        parser.error(f"Model path '{args.model}' does not exist.")

    if args.verbose:
        logger = Logger()
        logger.set_level(LogLevel.DEBUG)

    return args


if __name__ == "__main__":
    args = parse_args()
    logger = Logger()
    config = Configuration()
    config.set_enable(Configuration.ITEM.SHOW_MODEL_INFO, True)
    config.set_enable(Configuration.ITEM.SHOW_PROFILE, True)

    logger.info('Runtime framework version: ' + config.get_version())
    logger.info('Device driver version: ' + config.get_driver_version())
    logger.info('PCIe driver version: ' + config.get_pcie_driver_version())

    if config.get_enable(Configuration.ITEM.SHOW_MODEL_INFO):
        logger.info('SHOW_MODEL_INFO configuration is enabled')
    else:
        logger.info('SHOW_MODEL_INFO configuration is disabled')

    logger.info(f"Start run_async_model_conf test for model: {args.model}")
    result = -1

    try:

        # create inference engine instance with model
        with InferenceEngine(args.model) as ie:

            # register call back function
            ie.register_callback(on_inference_callback_func)

            # NOTE: np.zeros() uses COW zero pages — all virtual pages share one
            # physical page. PCIe DMA driver's get_user_pages() then sees duplicate
            # physical pages in the SG list and fails with EFAULT.
            # np.empty() + explicit fill forces unique physical page allocation.
            _buf = np.empty(ie.get_input_size(), dtype=np.uint8)
            _buf.fill(0)
            input = [_buf]
            start = time.perf_counter()

            # inference loop
            for i in range(args.loops):

                # inference asynchronously, use all npu cores
                # if device-load >= max-load-value, this function will block
                ie.run_async(input, user_arg=[i, args.loops])

                logger.debug(f"Inference start (async) {i}")

            # wait until all callback data processing is completed
            result = q.get()

            end = time.perf_counter()
            total_time_ms = (end -start) * 1000
            avg_latency = total_time_ms / args.loops
            fps = 1000.0/ avg_latency if avg_latency > 0 else 0.0

            logger.info("-----------------------------------")
            logger.info(f"Total Time: {total_time_ms:.3f} ms")
            logger.info(f"Average Latency: {avg_latency:.3f} ms")
            logger.info(f"FPS: {fps:.2f} frame/sec")
            logger.info("Success")
            logger.info("-----------------------------------")

            device_count = DeviceStatus.get_device_count()
            for i in range(device_count):
                device_status = DeviceStatus.get_current_status(i)
                logger.info(f'Device {device_status.get_id()}')

                for c in range(3):
                    logger.info(
                        f'   NPU Core {c} '
                        f'Temperature: {device_status.get_temperature(c)} '
                        f'Voltage: {device_status.get_npu_voltage(c)} '
                        f'Clock: {device_status.get_npu_clock(c)}'
                    )

    except Exception as e:
        logger.error(f"Exception: {str(e)}")
        exit(-1)

    exit(result)
